proposals/low-level-struct-improvements.md

```diff
@@ -849,24 +788,129 @@ escapes struct S
+
+The span safety rules for method invocation will be updated in several ways. The first is by recognizing the impact that `scoped` has on arguments. For a given argument `a` that is passed to parameter `p`:
+
+> 1. If `p` is `scoped ref` or `scoped in` then `a` does not contribute *ref-safe-to-escape* when considering arguments
+> 2. If `p` is `scoped` then `a` does not contribute *safe-to-escape* when considering arguments. Note that `ref scoped` is included here as the value is `scoped`
+
+The language "does not contribute" means the arguments are simply not considered when calculating the *ref-safe-to-escape* or *safe-to-escape* value of the method return respectively. That is because the values can't contribute to that lifetime (the `scoped` prevents it).
+
+The method invocation for lvalue returns can now be simplified. The receiver no longer needs to be special cased, in the case of `struct` it is now simply a `scoped ref T`, nor do `out` parameters need to be considered in the argument list. 
+
+> An lvalue resulting from a ref-returning method invocation `e1.M(e2, ...)` is *ref-safe-to-escape* the smallest of the following scopes:
+> 1. The *calling method*
+> 2. The *ref-safe-to-escape* of all `ref` arguments
+> 3. The *ref-safe-to-escape* of all `in` parameters when the argument is an lvalue otherwise *current method*
+> 4. The *safe-to-escape* of all argument expressions
+
+The method invocation for rvalue returns needs to change as follows to account for `ref` field returns.
+
+> An rvalue resulting from a method invocation `e1.M(e2, ...)` is *safe-to-escape* from the smallest of the following scopes:
+> 1. The *calling method*
+> 2. The *safe-to-escape* of all argument expressions
+> 3. When the return is a `ref struct` then *ref-safe-to-escape* of all `ref` and `in` arguments
+
+This rule now lets us define the two variants of desired methods:
+
+```c#
+Span<int> CreateWithoutCapture(scoped ref int value)
+{
+    // Error: RValue Rule 3 specifies that the safe-to-escape be limited to the ref-safe-to-escape
+    // of the ref argument. That is the current method for value hence this is not allowed.
+    return new Span<int>(ref value);
+}
+
+Span<int> CreateAndCapture(ref int value)
+{
+    // Okay: RValue Rule 3 specifies that the safe-to-escape be limited to the ref-safe-to-escape
+    // of the ref argument. That is the calling method for value hence this is not allowed.
+    return new Span<int>(ref value)
+}
+
+Span<int> ComplexScopedRefExample(scoped ref Span<int> span)
+{
+    // Okay: the safe-to-escape of `span` is *calling method* hence this is legal.
+    return span;
+
+    // Okay: the parameter here is `scoped ref` hence `refLocal` does not contribute /
```

Comment:
This example doesn't use the parameter?

Reply:
Correct.

Reply:
Then I don't understand the explanation. If `refLocal` isn't involved at all in this example, why is its scope or lack thereof relevant?

Reply:
`refLocal` is passed as an argument here so it's definitely involved. The point of this particular sample is to drive home that there are two lifetimes to consider here: the ref and the value. The parameter is `scoped ref` that means the ref lifetime does not contribute to the returned `ref struct` (the *ref-safe-to-escape* is not considered).

Compare that with the next sample where it concretely shows that the value does contribute here and prevent a return

Reply:
I think I'm confused because you say "parameter", which is ambiguous: did you mean `span` that is currently in scope, or did you mean the parameter of the called method (which is also this method, making it confusing to talk about)? Consider clarifying what parameter you mean.

Reply:
Is this explanation clearer? 

```
    // Okay: the local `refLocal` has a ref-safe-to-escape of *current method* and a 
    // safe-to-escape of *calling method*. In the call below it is passed to a 
    // parameter that is `scoped ref` which means it does not contribute 
    // ref-safe-to-escape. It only contributes its safe-to-escape hence the returned
    // rvalue ends up as safe-to-escape of *calling method*
```

Reply:
Much clearer, thanks!

Reply:
Awesome. Will push that change shortly